From 6133fc662cd33e31cea5f89e7f878fb835924aaf Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Tue, 12 Nov 2024 22:59:32 +0530
Subject: [PATCH 3/10] wifi: ath12k: add vendor support to process primary
 link migrate command

Introduce a new vendor sub-command
QCA_NL80211_VENDOR_SUBCMD_PRI_LINK_MIGRATE having value 0x100 (256) which
will be used to trigger primary link migration.

Add support to parse this new vendor NL sub-command. A subsequent change
will use the parsed values.

The format of iw command is -

$ iw dev wlan<X> vendor send 0x001374 0x100 <Link ID> <MAC ADDR>
<Link ID>: In hexadecimal so append with 0x
<MAC ADDR>: In hexadecimal so append with 0x

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/vendor.c | 43 ++++++++++++++++++++++++
 drivers/net/wireless/ath/ath12k/vendor.h |  2 +-
 2 files changed, 44 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/vendor.c
+++ b/drivers/net/wireless/ath/ath12k/vendor.c
@@ -4170,6 +4170,42 @@ int ath12k_vendor_put_ag_num_socs(struct
 	return 0;
 }
 
+static
+int ath12k_vendor_trigg_pri_link_migrate(struct wiphy *wiphy,
+					 struct wireless_dev *wdev,
+					 const void *data, int data_len)
+{
+	struct ieee80211_vif *vif = wdev_to_ieee80211_vif(wdev);
+	u8 mac_addr[ETH_ALEN] = {0};
+	u8 link_id;
+
+	if (WARN_ON(!vif))
+		return -EINVAL;
+
+	/* 1 byte of link ID or(and) 6 bytes of mac address */
+	if (data_len != 1 && data_len != ETH_ALEN + 1)
+		return -EINVAL;
+
+	/* not supported in case of non-ML vif */
+	if (!vif->valid_links)
+		return -EOPNOTSUPP;
+
+	/* get link ID */
+	link_id = *(u8 *)data;
+
+	/* get mac address if it is provided */
+	if (data_len == ETH_ALEN + 1) {
+		data++;
+		memcpy(mac_addr, data, ETH_ALEN);
+	}
+
+	ath12k_dbg(NULL, ATH12K_DBG_MAC,
+		   "primary link migration command received link_id %u, mac_addr %pM",
+		   link_id, mac_addr);
+
+	return 0;
+}
+
 static struct wiphy_vendor_command ath12k_vendor_commands[] = {
 	{
 		.info.vendor_id = QCA_NL80211_VENDOR_ID,
@@ -4261,6 +4297,14 @@ static struct wiphy_vendor_command ath12
 		.maxattr = QCA_WLAN_VENDOR_ATTR_CONFIG_MAX,
 		.flags = WIPHY_VENDOR_CMD_NEED_NETDEV,
 	},
+	{
+		.info.vendor_id = QCA_NL80211_VENDOR_ID,
+		.info.subcmd = QCA_NL80211_VENDOR_SUBCMD_PRI_LINK_MIGRATE,
+		.doit = ath12k_vendor_trigg_pri_link_migrate,
+		.policy = VENDOR_CMD_RAW_DATA,
+		.flags = WIPHY_VENDOR_CMD_NEED_NETDEV |
+			 WIPHY_VENDOR_CMD_NEED_RUNNING,
+	},
 };
 
 static const struct nl80211_vendor_cmd_info ath12k_vendor_events[] = {
--- a/drivers/net/wireless/ath/ath12k/vendor.h
+++ b/drivers/net/wireless/ath/ath12k/vendor.h
@@ -30,7 +30,7 @@ enum qca_nl80211_vendor_subcmds {
 	QCA_NL80211_VENDOR_SUBCMD_SET_6GHZ_POWER_MODE = 253,
 	QCA_NL80211_VENDOR_SUBCMD_POWER_MODE_CHANGE_COMPLETED = 254,
 	QCA_NL80211_VENDOR_SUBCMD_IQ_DATA_INFERENCING = 255,
-
+	QCA_NL80211_VENDOR_SUBCMD_PRI_LINK_MIGRATE = 256,
 	QCA_NL80211_VENDOR_SUBCMD_TELEMETRIC_DATA = 334,
 };
 
